File: nexago_app/lib/features/arenas/data/coupons_repository.dart
import 'package:cloud_functions/cloud_functions.dart';

import '../domain/arena_coupon.dart';

class CouponsException implements Exception {
  CouponsException(this.message);
  final String message;
  @override
  String toString() => message;
}

/// Gestão de cupons de marketing da arena (lado gestor) — callables em
/// `functions/src/arena-coupons.ts` (criar/listar/desativar). Diferente de
/// `PromotionsRepository`, que escreve direto no Firestore: cupom precisa
/// validar vigência/limite/XOR de desconto e checar resgates de forma
/// atômica, então toda escrita passa por Cloud Functions.
class CouponsRepository {
  CouponsRepository({FirebaseFunctions? functions})
      : _functions = functions ?? FirebaseFunctions.instance;

  final FirebaseFunctions _functions;

  Future<List<ArenaCoupon>> listCoupons(String arenaId) async {
    final id = arenaId.trim();
    if (id.isEmpty) return const [];
    try {
      final raw = await _functions
          .httpsCallable('listArenaCoupons')
          .call(<String, dynamic>{'arenaId': id});
      final data = raw.data;
      if (data is! Map) return const [];
      final list = data['coupons'];
      if (list is! List) return const [];
      return list
          .whereType<Map>()
          .map((m) => ArenaCoupon.fromMap(Map<String, dynamic>.from(m)))
          .toList();
    } on FirebaseFunctionsException catch (e) {
      throw CouponsException(_mapFunctionsMessage(e));
    }
  }

  /// Cria um cupom. `discountPercent` OU `fixedDiscountReais` (XOR, validado
  /// no servidor). `maxRedemptionsPerAthlete` default 1.
  Future<String> createCoupon({
    required String arenaId,
    required String code,
    double? discountPercent,
    double? fixedDiscountReais,
    DateTime? validFrom,
    DateTime? validUntil,
    int? maxRedemptionsTotal,
    int maxRedemptionsPerAthlete = 1,
  }) async {
    try {
      final payload = <String, dynamic>{
        'arenaId': arenaId.trim(),
        'code': code.trim(),
        'maxRedemptionsPerAthlete': maxRedemptionsPerAthlete,
        if (discountPercent != null) 'discountPercent': discountPercent,
        if (fixedDiscountReais != null)
          'fixedDiscountReais': fixedDiscountReais,
        if (validFrom != null) 'validFrom': validFrom.toIso8601String(),
        if (validUntil != null) 'validUntil': validUntil.toIso8601String(),
        if (maxRedemptionsTotal != null)
          'maxRedemptionsTotal': maxRedemptionsTotal,
      };
      final raw =
          await _functions.httpsCallable('createArenaCoupon').call(payload);
      final data = raw.data;
      if (data is! Map) {
        throw CouponsException('Resposta inválida do servidor.');
      }
      final couponId = (data['couponId'] as String?)?.trim() ?? '';
      if (couponId.isEmpty) {
        throw CouponsException('Não foi possível criar o cupom.');
      }
      return couponId;
    } on FirebaseFunctionsException catch (e) {
      throw CouponsException(_mapFunctionsMessage(e));
    }
  }

  /// Desativa o cupom (ação unidirecional — reativar exige criar outro código).
  Future<void> deactivateCoupon({
    required String arenaId,
    required String couponId,
  }) async {
    try {
      await _functions
          .httpsCallable('deactivateArenaCoupon')
          .call(<String, dynamic>{
        'arenaId': arenaId.trim(),
        'couponId': couponId.trim(),
      });
    } on FirebaseFunctionsException catch (e) {
      throw CouponsException(_mapFunctionsMessage(e));
    }
  }

  static String _mapFunctionsMessage(FirebaseFunctionsException e) {
    final detail = e.message;
    switch (e.code) {
      case 'unauthenticated':
        return 'Faça login para continuar.';
      case 'permission-denied':
        return detail ?? 'Você não gerencia esta arena.';
      case 'not-found':
        return detail ?? 'Não encontrado.';
      case 'already-exists':
        return detail ?? 'Já existe um cupom com este código.';
      case 'invalid-argument':
        return detail ?? 'Dados inválidos.';
      case 'failed-precondition':
        return detail ?? 'Não foi possível concluir.';
      case 'internal':
        return detail ?? 'Erro no servidor. Tente novamente.';
      default:
        return detail ?? 'Erro ao processar (${e.code}).';
    }
  }
}

File: frontend/projects/arena/src/app/painel/coupons/coupons-repository.ts
import { httpsCallable, type Functions } from 'firebase/functions';
import type { ArenaCoupon } from './coupon.model';

/** Diferente de `promotions-repository.ts` (escrita direta no Firestore): cupons passam por
 *  Cloud Functions callable (mesmo padrão de `arena-wallet-repository.ts`), porque criar/desativar
 *  exige revalidar código único e resgates em transação server-side. */

export class ArenaCouponError extends Error {}

function mapFunctionsError(err: unknown): ArenaCouponError {
  const message = err instanceof Error && err.message ? err.message : 'Não foi possível concluir a operação. Tente novamente.';
  return new ArenaCouponError(message);
}

export interface CreateArenaCouponInput {
  arenaId: string;
  code: string;
  discountPercent?: number | null;
  fixedDiscountReais?: number | null;
  /** ISO */
  validFrom?: string | null;
  /** ISO */
  validUntil?: string | null;
  maxRedemptionsTotal?: number | null;
  maxRedemptionsPerAthlete?: number;
}

export async function createArenaCoupon(functions: Functions, input: CreateArenaCouponInput): Promise<string> {
  const payload: Record<string, unknown> = {
    arenaId: input.arenaId,
    code: input.code,
    maxRedemptionsPerAthlete: input.maxRedemptionsPerAthlete ?? 1,
  };
  if (input.discountPercent != null) payload['discountPercent'] = input.discountPercent;
  if (input.fixedDiscountReais != null) payload['fixedDiscountReais'] = input.fixedDiscountReais;
  if (input.validFrom) payload['validFrom'] = input.validFrom;
  if (input.validUntil) payload['validUntil'] = input.validUntil;
  if (input.maxRedemptionsTotal != null) payload['maxRedemptionsTotal'] = input.maxRedemptionsTotal;

  try {
    const result = await httpsCallable<Record<string, unknown>, { couponId: string }>(functions, 'createArenaCoupon')(payload);
    return result.data.couponId;
  } catch (err) {
    throw mapFunctionsError(err);
  }
}

export async function listArenaCoupons(functions: Functions, arenaId: string): Promise<ArenaCoupon[]> {
  try {
    const result = await httpsCallable<{ arenaId: string }, { coupons: ArenaCoupon[] }>(functions, 'listArenaCoupons')({ arenaId });
    return result.data.coupons ?? [];
  } catch (err) {
    throw mapFunctionsError(err);
  }
}

/** Irreversível — reativar exige criar outro código. */
export async function deactivateArenaCoupon(functions: Functions, arenaId: string, couponId: string): Promise<void> {
  try {
    await httpsCallable<{ arenaId: string; couponId: string }, { success: boolean }>(functions, 'deactivateArenaCoupon')({
      arenaId,
      couponId,
    });
  } catch (err) {
    throw mapFunctionsError(err);
  }
}
